In [2]:
import tensorflow as tf
from tensorflow.keras import layers, models
import numpy as np

In [3]:

model = models.Sequential()
model.add(layers.Conv2D (32, kernel_size=(3,3), padding='valid', activation='relu',

input_shape=(28,28,1)))
model.add(layers.Conv2D (32, kernel_size=(3,3), padding='valid', activation='relu'))
model.add(layers.Conv2D (32, kernel_size=(3,3), padding="valid", activation='relu'))
model.add(layers.Flatten())
model.add(layers.Dense (128, activation='relu'))
model.add(layers.Dense (10, activation='softmax'))

/Users/brianrono/anaconda3/envs/dev_env_clean/lib/python3.11/site-packages/keras/src/layers/convolutional/base_conv.py:113: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [4]:
model.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv2d (Conv2D)                 │ (None, 26, 26, 32)     │           320 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_1 (Conv2D)               │ (None, 24, 24, 32)     │         9,248 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_2 (Conv2D)               │ (None, 22, 22, 32)     │         9,248 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten (Flatten)               │ (None, 15488)          │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ (None, 128)            │     1,982,592 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 10)             │         1,290 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 2,002,698 (7.64 MB)

 Trainable params: 2,002,698 (7.64 MB)

 Non-trainable params: 0 (0.00 B)

Every layer reduces the spatial size by 2 pixels per convolution (because 3×3 kernel without padding removes 1 pixel from each edge).

Effect

*   Image shrinks → 28 → 26 → 24 → 22

*   Flattened size: smaller (15,488)
**bold text**
*   Parameters: ~2M

In [5]:
model = models.Sequential()
model.add(layers.Conv2D (32, kernel_size=(3,3), padding='same', activation='relu',
input_shape=(28,28,1)))
model.add(layers.Conv2D (32, kernel_size=(3,3), padding='same', activation='relu'))
model.add(layers.Conv2D (32, kernel_size=(3,3), padding='same', activation='relu'))
model.add(layers.Flatten())
model.add(layers.Dense (128, activation='relu'))
model.add(layers.Dense (10, activation='softmax'))

In [6]:
model.summary()

Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv2d_3 (Conv2D)               │ (None, 28, 28, 32)     │           320 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_4 (Conv2D)               │ (None, 28, 28, 32)     │         9,248 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_5 (Conv2D)               │ (None, 28, 28, 32)     │         9,248 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten_1 (Flatten)             │ (None, 25088)          │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 128)            │     3,211,392 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_3 (Dense)                 │ (None, 10)             │         1,290 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 3,231,498 (12.33 MB)

 Trainable params: 3,231,498 (12.33 MB)

 Non-trainable params: 0 (0.00 B)

Each layer preserves the same spatial dimension, because padding compensates for the border loss.

Effect

*   Image size stays constant → 28×28 all through.

*   Flattened size is larger (25,088).

*   Parameters grow to 3.2M, since the dense layer receives a bigger input.

In [7]:
model = models.Sequential()
model.add(layers.Conv2D (32, kernel_size=(3,3), padding='same', strides=(4,4), activation='relu',
input_shape=(28,28,1)))
model.add(layers.Conv2D (32, kernel_size=(3,3), padding='same', strides=(4,4), activation='relu'))
model.add(layers.Conv2D (32, kernel_size=(3,3), padding='same', strides=(4,4), activation='relu'))
model.add(layers.Flatten())
model.add(layers.Dense (128, activation='relu'))
model.add(layers.Dense (10, activation='softmax'))

In [8]:
model.summary()

Model: "sequential_2"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv2d_6 (Conv2D)               │ (None, 7, 7, 32)       │           320 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_7 (Conv2D)               │ (None, 2, 2, 32)       │         9,248 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_8 (Conv2D)               │ (None, 1, 1, 32)       │         9,248 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten_2 (Flatten)             │ (None, 32)             │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_4 (Dense)                 │ (None, 128)            │         4,224 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_5 (Dense)                 │ (None, 10)             │         1,290 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 24,330 (95.04 KB)

 Trainable params: 24,330 (95.04 KB)

 Non-trainable params: 0 (0.00 B)

Each convolution now halves the resolution because stride = 2 causes the kernel to “jump” 2 pixels at a time.

Effect

*   Rapid downsampling: 28 → 14 → 7 → 4

*   Flattened size becomes very small (512).

*   Parameters drop drastically (~85k).

*   Model is lighter and faster, but captures coarser features.